harness/determined/__init__.py

```diff
@@ -2,19 +2,17 @@
 from determined._env_context import EnvContext
 from determined._rendezvous_info import RendezvousInfo
 from determined._execution import (
-    _catch_init_invalid_hp,
     _catch_sys_exit,
     _make_local_execution_env,
     _local_execution_manager,
     InvalidHP,
 )
 from determined._train_context import NativeContext, TrialContext, DistributedContext, RankInfo
+from determined._generic import _GenericContext
```

Comment:
TL;DR I hadn't fully thought through where they should go when I wrote that, and while they are destined for being public, they're not public yet.

Having thought more about it, I think I'd like the eventual file layout to be:

```
determined/
deterimned/generic/
determined/generic/_context.py
determined/generic/_training.py
determined/generic/_searcher.py
determined/generic/_preemption.py
determined/generic/_checkpointing.py
```

and for the symbols to be:
```
det.generic.GenericContext
det.generic.Training
det.generic.SearcherOp
det.generic.AdvancedSearcher
det.generic.Preemption
det.generic.Checkpointing
```

and a flattened training loop using the Generic API could do something like:
```python
import determined as det
from determined import generic

# Generic + Dtrain requires an appropriate DistributedContext:
dist = det.DistributedContext(chief_addr=..., rank_info=...)

# GenericContext would take the same args as a Session,
# plus a DistributedContext (during dtrain anyway).
with generic.GenericContext(master_url=..., dist=dist) as ctx:
    epoch = 0
    batch = 0
    for op in det.searcher.ops():
        while epoch < op.epochs:
            if ctx.preemption.should_preempt():
                exit(0)
            total_batches += do_train_epoch()
            epoch += 1
            ctx.report_training_metrics(...)
            metrics = do_validate()
            ctx.report_validation_metrics(...)
```

I might also expose the `Checkpointing` methods and the `Preemption` methods right on the `GenericContext` since they should be exposed for every kind of Task, training or otherwise.

If this layout seems good to you, I can make the change now.

Reply:
The file layout looks good.
But if `GenericContext` is intended for public consumption, and it's a popular use case, then yes it should be additionally exposed on `det` when it's ready.

My "vision" for canonical `det` is:
- top 10/top 20 popular user-facing APIs should be in `det`.
- it's okay to alias deep symbols like `determined.nested1.nested2.Symbol` in `det`.
- no internal "convenience" imports
- no private-ish underscore stuff.